src/recycle/core.clj

```diff
@@ -0,0 +1,184 @@
+  [v]
+  (::service v))
+
+(defn service-factory?
+  [v]
+  (::factory v))
+
+;; --- Main API
+
+(def ^:dynamic *timeout*
+  "Maximum number of milliseconds to wait for a response from a
+  service. This variable by default is nil, meaning the value used
+  when creating the service is used."
+  nil)
+
+(defn- default-receive
+  "A default implementation for `:receive`."
+  [& _]
+  (ex-info "service do not implement :receive" {}))
+
+(declare initialize-service)
+(declare start!)
+
+(defn service
+  "Create a service using the spec map as a blueprint.
+  The spec map may contain any of the following keys (all of the are
+  optional):
+
+   - `:init`:       initialization hook, receives the options object passed to start
+                    function; it should create and return the local state of the service;
+                    (this attr is optional)
+   - `:stop`:       termination hook, receives the service spec, internal state and
+                    allows proper resource clearing; this parameter is optional.
+   - `:receive`:    function that receives the internal state and variable number of args
+                    identifying the message sent to the service.
+   - `:timeout`:    default maximum number of milliseconds to wait for a response
+                    from the service (default 1min).
+   - `:buf-or-n`:   core.async buffer or size of buffer to use for the
+                    service communication channel (default 1024).
+  "
+  [spec]
+  (letfn [(factory [{:keys [::instances] :as options}]
+            (let [options (dissoc options ::instances)]
+              (-> (assoc spec :options options)
+                  (merge (when instances {:instances instances}))
+                  (initialize-service))))]
+    {::factory factory}))
+
+(defn create
+  "Create an instance of the provided service."
+  ([sf]
+   (create sf nil))
+  ([sf options]
+   {:pre [(service-factory? sf)]}
+   ((::factory sf) options)))
+
+(defn started?
+  "Check if service is started. Returns true if it is, false if it's not."
+  [{:keys [::status] :as service}]
+  {:pre [(service? service)]}
+  (= @status ::started))
+
+(defn stopped?
+  "Check if service is stopped. Returns true if it is, false if it's not."
+  [{:keys [::status] :as service}]
+  {:pre [(service? service)]}
+  (= @status ::stopped))
+
+(declare initialize-loop)
+
+(defn start!
+  [{:keys [::status ::stop-ch ::local ::init ::options] :as service}]
+  {:pre [(service? service)]}
+  (when (compare-and-set! status ::stopped ::started)
+    (vreset! local (init options))
+    (vreset! stop-ch (a/chan))
+    (initialize-loop service)
+    service))
+
+(defn stop!
+  [{:keys [::status ::local ::stop-ch ::stop] :as service}]
+  {:pre [(service? service)]}
+  (when (compare-and-set! status ::started ::stopped)
+    (a/close! @stop-ch) ;; Notify the internal loop that the service is stoped
+    (stop @local)
+    (vreset! stop-ch nil)
+    (vreset! local nil)))
+
+(defn with-state
```

Comment:
I'm not completely sure about this 🤔 I think I would like more a higher order function that receives the transformation, something like this?:

```clj
:receive (with-state 
                (fn [counter message]
                  (a/go
                     (a/<! (a/timeout 10)) ;; simulate some async work
                     (inc counter)))}))
```